### PR TITLE
enhancement(journald source): Add timeout to batch read loop

### LINKED ISSUE #11187: `journald` source needs a `batch.timeout` option
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Current Vector Version

```
vector 0.19.1 (x86_64-unknown-linux-gnu 3cf70cf 2022-01-25)
```

### Use-cases

A high throughput `journald` source with `acknowledgements.enabled` when paired with a sink that batches output.

### Attempted Solutions

An example config:
```
[sources.journald]
    type = "journald"
    acknowledgements.enabled = true
    batch_size = 500

[sinks.firehose]
    type = "aws_kinesis_firehose"
    inputs = [ "journald" ]
    stream_name = "logs"
    encoding.codec = "json"
    ...
```

This works well to fill the default batch of the `firehose` sink (500) assuming there are at least 500 events in the `journald` source. Once the backlog in the `journald` source is consumed the source appears to block until it has met the `batch_size` threshold.

### Proposal

Use the same `batch.max_events` and `batch.timeout` pattern that sinks use in sources that can batch. This would allow for quick consumption of a backlog and keeping up with a high logging host while not penalizing slow logging periods too much.

### References

#9856

### PR DESCRIPTION
**Pull request by bruceg**
This adds a hard-coded timeout to the process of reading batches from the `journalctl` pipeline. There is a lot of code reorganization here mostly to avoid duplicating the logic inside the `run_stream` loop, but the core logic change is fairly simple: read one line, start a timeout, and then read lines until either that timeout or the batch size is reached.

Note that there are currently no integration tests, and hence no tests of this feature as it requires a running `journalctl` subprocess. Ref #11672

Closes #11187

### CODE REVIEW

File: src/sources/journald.rs [hunk 1]
```diff
@@ -302,103 +286,138 @@ impl JournaldSource {
...
-                            %error,
-                        );
+            let mut batch = Batch::new(self, checkpointer);
+
+            if !batch.handle_next(stream.next().await) {
+                break true;
+            }
+
+            let timeout = tokio::time::sleep(BATCH_TIMEOUT);
+            tokio::pin!(timeout);
```
Comment (blt) [inline src/sources/journald.rs:300]: I suppose an alternate approach might be to have an [`Interval`](https://docs.rs/tokio/latest/tokio/time/fn.interval.html) in the struct itself and poll that for ticks, although it wouldn't be lined up neatly with batches like this is.

**Review by blt — APPROVED** [review]
This makes sense to me. I had one stray thought that floated to the surface while reviewing but that's all. Tidy.

File: src/sources/journald.rs (hunk 1 above)
Comment (bruceg) [reply src/sources/journald.rs:300]: I think that would be unwelcome for sites where journald doesn't actually emit a lot of logs, and unnecessary for sites where it does.

**Review by jszwedko — APPROVED** [review]
Looks good. Thanks for breaking up the commits. That made reviewing easier.

**Pull request merged**